File: bnf/Python3/bnfParserBase.py
from antlr4 import *

relativeImport = False
if __name__ is not None and "." in __name__:
    relativeImport = True

class bnfParserBase(Parser):
    @staticmethod
    def parser():
        if relativeImport:
            from .bnfParser import bnfParser
        else:
            from bnfParser import bnfParser
        return bnfParser

    def NotNL(self):
        bnfParser = self.parser()
        i = 1
        c = self._input.LT(i)
        return c.type != bnfParser.NL

    def not_assign(self):
        i = 1
        c = self._input.LT(i)
        while c is not None and c.type == bnfLexer.WS:
            i += 1
            c = self._input.LT(i)
        return not (c.type == bnfLexer.ASSIGN1 or
                    c.type == bnfLexer.ASSIGN2 or
                    c.type == bnfLexer.ASSIGN3 or
                    c.type == bnfLexer.ASSIGN4)

    def not_lhs(self):
        i = 1
        c = self._input.LT(i)
        while c is not None and c.type == bnfLexer.WS:
            i += 1
            c = self._input.LT(i)
        if c is not None and c.type != bnfLexer.X1:
            return True
        # '<'
        while True:
            while c is not None and c.type == bnfLexer.WS:
                i += 1
                c = self._input.LT(i)
            if c is None or c.type not in (bnfLexer.ID, bnfLexer.X2):
                return True
            # ID
            if c.type == bnfLexer.X2:
                break
        # '>'
        while c is not None and c.type == bnfLexer.WS:
            i += 1
            c = self._input.LT(i)
        return not (c.type in (bnfLexer.ASSIGN1, bnfLexer.ASSIGN2, bnfLexer.ASSIGN3, bnfLexer.ASSIGN4))

        

File: bnf/CSharp/bnfParserBase.cs
using Antlr4.Runtime;
using System.IO;

public abstract class bnfParserBase : Parser {
    private readonly ITokenStream _input;

    protected bnfParserBase(ITokenStream input, TextWriter output, TextWriter errorOutput)
            : base(input, output, errorOutput)
    {
        _input = input;
    }

    public bool NotNL()
    {
        int i = 1;
        IToken c = (this.InputStream as CommonTokenStream).LT(i);
        var v = c.Type != bnfParser.NL;
        return v;
    }

    public bool NotAssign()
    {
        int i = 1;
        IToken c = (this.InputStream as CommonTokenStream).LT(i);
        while (c != null && c.Type == bnfLexer.WS)
            c = (this.InputStream as CommonTokenStream).LT(++i);
        var v = !(c.Type == bnfParser.ASSIGN1
             | c.Type == bnfParser.ASSIGN2
             | c.Type == bnfParser.ASSIGN3
             | c.Type == bnfParser.ASSIGN4);
        return v;
    }

    public bool NotLhs()
    {
        int i = 1;
        IToken c = (this.InputStream as CommonTokenStream).LT(i);
        while (c != null && c.Type == bnfLexer.WS)
            c = (this.InputStream as CommonTokenStream).LT(++i);
        if (c != null && c.Type != bnfLexer.X1)
            return true;
        // '<'
        for (;;)
        {
            while (c != null && c.Type == bnfLexer.WS)
                c = (this.InputStream as CommonTokenStream).LT(++i);
            if (c != null && c.Type != bnfLexer.ID && c.Type != bnfLexer.X2)
                return true;
            // ID
            if (c == null) return true;
            if (c.Type == bnfLexer.X2)
                break;
        }
        // '>'
        while (c != null && c.Type == bnfLexer.WS)
            c = (this.InputStream as CommonTokenStream).LT(++i);
        if (c.Type == bnfLexer.ASSIGN1) return false;
        if (c.Type == bnfLexer.ASSIGN2) return false;
        if (c.Type == bnfLexer.ASSIGN3) return false;
        if (c.Type == bnfLexer.ASSIGN4) return false;
        return true;
    }

    
}